Check ashift validity in 'zpool add'

loli10K opened the pull request:
### Description
<!--- Describe your changes in detail -->

df83110 added the ability to specify a custom "ashift" value from the command line in 'zpool add' and 'zpool attach'. This commit adds additional checks to the provided ashift to prevent invalid values from being used, which could result in disastrous consequences (https://github.com/zfsonlinux/zfs/issues/4434) for the whole pool.

More context: reading the following comments i understand both `zpool_create` and `zpool_add` expect the "nvroot" parameter to be valid:

```c
/*
 * Create the named pool, using the provided vdev list.  It is assumed
 * that the consumer has already validated the contents of the nvlist, so we
 * don't have to worry about error semantics.
 */
zpool_create(libzfs_handle_t *hdl, const char *pool, nvlist_t *nvroot,
/*
 * Add the given vdevs to the pool.  The caller must have already performed the
 * necessary verification to ensure that the vdev specification is well-formed.
 */
zpool_add(zpool_handle_t *zhp, nvlist_t *nvroot)
```

![zcallgraph](https://cloud.githubusercontent.com/assets/4585738/23715963/cbfac990-042e-11e7-93ef-085f340749c9.png)

This seems to be false in both cases since the ashift value is parsed in `make_leaf_vdev()` by `zfs_nicestrtonum()`, but is never explicitly validated by `zpool_valid_proplist()`:



```
root@debian-8-zfs:~# cat > gdbcmds <<EOF
> set pagination off
> set breakpoint pending on
> b cmd/zpool/zpool_vdev.c:720
> file zpool
> run add $POOLNAME -o ashift=1e $TMPDIR/zpool2.dat
> EOF
root@debian-8-zfs:~# gdb -q -x gdbcmds
No symbol table is loaded.  Use the "file" command.
Breakpoint 1 (cmd/zpool/zpool_vdev.c:720) pending.
[Thread debugging using libthread_db enabled]
Using host libthread_db library "/lib/x86_64-linux-gnu/libthread_db.so.1".

Breakpoint 1, make_leaf_vdev (props=props@entry=0x625ea0, arg=<optimized out>, is_log=is_log@entry=0) at zpool_vdev.c:720
720		if (props != NULL) {
(gdb) bt
#0  make_leaf_vdev (props=props@entry=0x625ea0, arg=<optimized out>, is_log=is_log@entry=0) at zpool_vdev.c:720
#1  0x0000000000414a76 in construct_spec (props=0x625ea0, argc=argc@entry=1, argv=argv@entry=0x7fffffffecc0) at zpool_vdev.c:1592
#2  0x00000000004150b6 in make_root_vdev (zhp=0x625f00, props=<optimized out>, force=0, check_rep=1, replacing=B_FALSE, dryrun=B_FALSE, argc=1, argv=0x7fffffffecc0) at zpool_vdev.c:1721
#3  0x000000000040e39e in zpool_do_add (argc=1, argv=0x625f00) at zpool_main.c:680
#4  0x00000000004056ad in main (argc=6, argv=0x7fffffffec98) at zpool_main.c:7510
(gdb) b zfs_nicestrtonum
Breakpoint 2 at 0x7ffff722c6e0: file libzfs_util.c, line 1540.
(gdb) c
Continuing.

Breakpoint 2, zfs_nicestrtonum (hdl=hdl@entry=0x0, value=0x625ee8 "1e", num=num@entry=0x7fffffff9828) at libzfs_util.c:1540
1540	{
(gdb) fin
Run till exit from #0  zfs_nicestrtonum (hdl=hdl@entry=0x0, value=0x625ee8 "1e", num=num@entry=0x7fffffff9828) at libzfs_util.c:1540
0x0000000000414534 in make_leaf_vdev (props=props@entry=0x625ea0, arg=<optimized out>, is_log=is_log@entry=0) at zpool_vdev.c:725
725				zfs_nicestrtonum(NULL, value, &ashift);
Value returned is $1 = 0
(gdb) p ashift
$2 = 1152921504606846976
(gdb) 
```

"ashift=1e" (an understandable typo) is parsed as 1 exabyte (1152921504606846976). `zpool` drives on, adds the new file vdev and then panics the box:

```
(gdb) p ashift
$2 = 1152921504606846976
(gdb) c
Continuing.
[Inferior 1 (process 1971) exited normally]
(gdb) q
root@debian-8-zfs:~# zdb -C $POOLNAME | grep ashift:
                ashift: 9
                ashift: 1152921504606846976
root@debian-8-zfs:~# [  520.750214] KGDB: Waiting for remote debugger
```

From the kernel debugger:

```
Thread 119 received signal SIGSEGV, Segmentation fault.
[Switching to Thread 1779]
0xffffffffc0408a35 in metaslab_group_histogram_remove (mg=0xffff880029942c00, msp=0xffff88001814f400) at /usr/src/zfs/module/zfs/metaslab.c:765
765			mg->mg_histogram[i + ashift] -=
(gdb) bt
#0  0xffffffffc0408a35 in metaslab_group_histogram_remove (mg=0xffff880029942c00, msp=0xffff88001814f400) at /usr/src/zfs/module/zfs/metaslab.c:765
#1  0xffffffffc0409664 in metaslab_sync (msp=0xffff88001814f400, txg=20) at /usr/src/zfs/module/zfs/metaslab.c:2259
#2  0xffffffffc042bbbd in vdev_sync (vd=0xffff88002a7c8000, txg=20) at /usr/src/zfs/module/zfs/vdev.c:2399
#3  0xffffffffc04166bb in spa_sync (spa=0xffff8800181b6000, txg=20) at /usr/src/zfs/module/zfs/spa.c:6624
#4  0xffffffffc0427111 in txg_sync_thread (dp=0xffff8800181f5000) at /usr/src/zfs/module/zfs/txg.c:545
#5  0xffffffffc02b84ae in thread_generic_wrapper ()
#6  0xffffffff8109ac4f in kthread ()
#7  0xffffffff815cc3f2 in ret_from_fork () at arch/x86/entry/entry_64.S:406
#8  0x0000000000000000 in ?? ()
(gdb) list
760			ASSERT3U(mg->mg_histogram[i + ashift], >=,
761			    msp->ms_sm->sm_phys->smp_histogram[i]);
762			ASSERT3U(mc->mc_histogram[i + ashift], >=,
763			    msp->ms_sm->sm_phys->smp_histogram[i]);
764	
765			mg->mg_histogram[i + ashift] -=
766			    msp->ms_sm->sm_phys->smp_histogram[i];
767			mc->mc_histogram[i + ashift] -=
768			    msp->ms_sm->sm_phys->smp_histogram[i];
769		}
(gdb) p ashift
$1 = 1152921504606846976
```

We are lucky `zpool_create()` detects the invalid ashift with an explicit call to `zpool_valid_proplist()`, even if "nvroot" already contains the invalid value.

Unfortunately `zpool_add()` does not call `zpool_valid_proplist()` (it doesn't even have a "props" argument to begin with).

How do we fix this:

1. pass "props" to `zpool_add()`, validate them with `zpool_valid_proplist()`.
2. validate "ashift" in `make_leaf_vdev()`
3. ???


### Motivation and Context
Fix https://github.com/zfsonlinux/zfs/issues/4434

### How Has This Been Tested?
Test case is provided. Also manual debugging.

### Types of changes
<!--- What types of changes does your code introduce? Put an `x` in all the boxes that apply: -->
- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Performance enhancement (non-breaking change which improves efficiency)
- [ ] Code cleanup (non-breaking change which makes code smaller or more readable)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

### Checklist:
<!--- Go over all the following points, and put an `x` in all the boxes that apply. -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
- [x] My code follows the ZFS on Linux code style requirements.
- [ ] I have updated the documentation accordingly.
- [x] I have read the **CONTRIBUTING** document.
- [x] I have added tests to cover my changes.
- [ ] All new and existing tests passed.
- [ ] Change has been approved by a ZFS on Linux member.

dinatale2 commented on tests/zfs-tests/tests/functional/cli_root/zpool_add/add-o_ashift.ksh:
```diff
@@ -0,0 +1,109 @@
+
+typeset ashifts=("9" "10" "11" "12" "13" "14" "15" "16")
+typeset -i i=0;
+while [ $i -lt "${#ashifts[@]}" ]
```
nit: a `for` loop may be cleaner here.

dinatale2 commented on tests/zfs-tests/tests/functional/cli_root/zpool_add/add-o_ashift.ksh:
```diff
@@ -0,0 +1,109 @@
+
+typeset badvals=("off" "on" "1" "8" "17" "1b" "ff" "-")
+typeset -i i=0;
+while [ $i -lt "${#badvals[@]}" ]
```
nit: a `for` loop may be cleaner here.

Also, @behlendorf should this be broken apart into a pos and neg test cases?

dinatale2 commented on cmd/zpool/zpool_vdev.c:
```diff
@@ -721,8 +721,15 @@ make_leaf_vdev(nvlist_t *props, const char *arg, uint64_t is_log)
-		    zpool_prop_to_name(ZPOOL_PROP_ASHIFT), &value) == 0)
+		    zpool_prop_to_name(ZPOOL_PROP_ASHIFT), &value) == 0) {
 			zfs_nicestrtonum(NULL, value, &ashift);
+			if (ashift != 0 && (ashift < 9 || ashift > 16)) {
```
I didn't see them defined anywhere, but would it be worth having `ASHIFT_MIN` and `ASHIFT_MAX` defined in an appropriate header file? Thoughts?

loli10K replied:
Agreed. The same header file would also be needed in "lib/libzfs/libzfs_pool.c" (`zpool_valid_proplist()`) and "module/zfs/vdev.c" (`vdev_open()`): right now i don't see anything already shared among the 3.

dinatale2 commented:
@loli10K Forgot to ask in my review, did you intend to put `Requires-builders: none` in your commit message? The buildbot won't run anything with that option specified in the commit message.

loli10K commented:
> did you intend to put Requires-builders: none in your commit message?

Yes, no need to bother the BB and waste resources on the first take, i just wanted to propose a fix and get a review: i'm going to drop it from the commit message once the code looks good.

dinatale2 commented:
For your build issues, does something like `fprintf("%" PRIu64 "\n", val);` not work?

loli10K commented:
@dinatale2 the last push should have fixed the build issue, i borrowed the code from zdb: it seems the BB is still building the very first commit, maybe it's already time to squash this thing?

The other thing that's still missing is a comment in "sys/spa.h" to explain ASHIFT_MIN and MAX definitions.

dinatale2 commented:
@loli10K You should rebase on master when you get a change to get rid of the style issues. They were resolved yesterday.

loli10K commented:
I ended up hammering the BB anyway ... orz 

This PR was accidentally bumping the ashift value to 16: this works should be done in the other PR (https://github.com/zfsonlinux/zfs/pull/5763), so i lowered it back to 13 in "sys/spa.h".

dinatale2 commented:
Looks like the STYLE builder was having trouble this morning. I'm working on resolving it. I'll resubmit your last build to the STYLE builder once I've resolved it.

dinatale2 commented on module/zfs/vdev.c:
```diff
@@ -1347,6 +1347,10 @@ vdev_open(vdev_t *vd)
 			vd->vdev_ashift = ashift;
+		else
+			if (vd->vdev_ashift < ASHIFT_MIN ||
+			    vd->vdev_ashift > ASHIFT_MAX)
```
At this point, if the ashift is invalid, does that mean the vdev can't be opened? If that is the case, should the vdev have its state set to `VDEV_STATE_CANT_OPEN` with something like `VDEV_AUX_BAD_ASHIFT`?

loli10K replied:
Yes, i'm going to update the code as soon as i get back to my dev machine. Also, am i allowed to add `VDEV_AUX_BAD_ASHIFT` to the `vdev_aux_t` enum or should i use an existing value?

dinatale2 replied:
You can add `VDEV_AUX_BAD_ASHIFT` as an entry at the end of the `vdev_aux_t` enum. This should result in the disk being labelled as `UNAVAIL` which I think is appropriate. The fact that this case also returns `EDOM` provides even more logging. I would still check upstream to make sure the enum hasn't changed as well, but we've pulled in almost all of the upstream patches and it hasn't changed thus far.

@behlendorf any other thoughts?

behlendorf replied:
Let's add a new value and then make sure this gets pushed upstream so we stay in sync.    In general the more information we can provide about why the open failed the better.  We're already doubling up on some of the VEDV_AUX errors which can be confusing when debugging.

loli10K replied:
`vdev_aux_t` hasn't been changed in ~5 years (https://github.com/openzfs/openzfs/blame/master/usr/src/uts/common/sys/fs/zfs.h#L643), i think it's safe to add `VDEV_AUX_BAD_ASHIFT` to it.

I also tried to add just enough code so when we bubble up the error from `vdev_open()` we have sufficient information to generate a new "vdev.bad_ashift" FMA event (instead of a generic "vdev.unknown"):

```
root@debian-8-zfs:~# fallocate -l 64m $TMPDIR/zpool1.dat
root@debian-8-zfs:~# fallocate -l 64m $TMPDIR/zpool2.dat
root@debian-8-zfs:~# zpool create -f $POOLNAME $TMPDIR/zpool1.dat
root@debian-8-zfs:~# zpool events -c
cleared 8 events
root@debian-8-zfs:~# zpool add $POOLNAME -o ashift=14 $TMPDIR/zpool2.dat # <- without userland ashift checks
cannot add to 'testpool': block size out of range or does not match
root@debian-8-zfs:~# zpool events -v
TIME                           CLASS
Mar 14 2017 19:42:41.707571434 ereport.fs.zfs.vdev.bad_ashift
        class = "ereport.fs.zfs.vdev.bad_ashift"
        ena = 0x21942bacc800001
        detector = (embedded nvlist)
                version = 0x0
                scheme = "zfs"
                pool = 0xd4f8c6bab1189a81
                vdev = 0xcdb757669057238d
        (end detector)
        pool = "testpool"
        pool_guid = 0xd4f8c6bab1189a81
        pool_context = 0x0
        pool_failmode = "wait"
        vdev_guid = 0xcdb757669057238d
        vdev_type = "file"
        vdev_path = "/var/tmp/zpool2.dat"
        vdev_ashift = 0xe
        vdev_complete_ts = 0x0
        vdev_delta_ts = 0x0
        vdev_read_errors = 0x0
        vdev_write_errors = 0x0
        vdev_cksum_errors = 0x0
        parent_guid = 0x545994fec1996fa3
        parent_type = "root"
        vdev_spare_paths = 
        vdev_spare_guids = 
        prev_state = 0x7
        time = 0x58c839a1 0x2a2caeea 
        eid = 0x10

Mar 14 2017 19:42:41.707571434 resource.fs.zfs.statechange
        version = 0x0
        class = "resource.fs.zfs.statechange"
        pool_guid = 0xd4f8c6bab1189a81
        pool_context = 0x0
        vdev_guid = 0xcdb757669057238d
        vdev_state = "UNAVAIL" (0x4)
        vdev_path = "/var/tmp/zpool2.dat"
        vdev_laststate = "ONLINE" (0x7)
        time = 0x58c839a1 0x2a2caeea 
        eid = 0x11
```

dinatale2 commented on tests/zfs-tests/tests/functional/cli_root/zpool_add/add-o_ashift.ksh:
```diff
@@ -0,0 +1,104 @@
+#
+# STRATEGY:
+#	1. Create a pool with default values.
+#	2. Verify 'zpool add -o ashift=<n>' works with allowed values (9-16).
```
nit: Testing a sampling of bad values is also part of your strategy. You're also testing it with values from 9-13.

dinatale2 commented on include/sys/spa.h:
```diff
@@ -122,6 +122,17 @@ _NOTE(CONSTCOND) } while (0)
 
 /*
+ * Alignment Shift (ashift) is an immutable, internal top-level vdev property
+ * which can only be set at creation time. Physical writes are always done
```
nit: Should this say `at vdev creation time`?

dinatale2 commented on tests/zfs-tests/tests/functional/cli_root/zpool_add/add-o_ashift.ksh:
```diff
@@ -0,0 +1,104 @@
+	log_must $ZPOOL create $TESTPOOL $disk1
+	log_mustnot $ZPOOL add -o ashift="$badval" $disk2
+	log_must $ZPOOL destroy $TESTPOOL
+	log_must $ZPOOL labelclear $disk1
```
Shouldn't we do a label clear on `$disk2` just to be safe?

loli10K replied:
You're right, i thought `zpool labelclear` would return an error if it can't find labels on the specified device (and since we expect "_$ZPOOL add -o ashift="$badval" $disk2_" to fail we should also expect no labels on `$disk2`), but i just checked and that's not the case: i'm going to add the other label clear.

loli10K commented:
@dinatale2 it's my fault, i think i added a minimal runfile with only the new testcase but that got lost when i squashed the commits and restored the default `TEST` file:

```
root@debian-8-zfs:~# ls -l /usr/share/zfs/runfiles/
total 40
-rwxr-xr-x 1 root root   766 Mar  6 11:04 issue-1350.run
-rwxr-xr-x 1 root root   612 Mar  8 15:41 issue-4434.run
-rwxr-xr-x 1 root root 25127 Mar 14 16:03 linux.run
-rwxr-xr-x 1 root root   807 Mar 14 16:03 perf-regression.run
```

I'll update the main runfile in a few hours orz.

behlendorf commented on cmd/zpool/zpool_vdev.c:
```diff
@@ -721,8 +722,16 @@ make_leaf_vdev(nvlist_t *props, const char *arg, uint64_t is_log)
 			zfs_nicestrtonum(NULL, value, &ashift);
+			if (ashift != 0 &&
+			    (ashift < ASHIFT_MIN || ashift > ASHIFT_MAX)) {
+				(void) fprintf(stderr, dgettext(TEXT_DOMAIN,
```
Since this is in the command and not a library you'll want to use `gettext()` for consistency with the errors above.  The text domain is already set to TEXT_DOMAIN.

behlendorf commented on cmd/zpool/zpool_vdev.c:
```diff
@@ -721,8 +722,16 @@ make_leaf_vdev(nvlist_t *props, const char *arg, uint64_t is_log)
+			    (ashift < ASHIFT_MIN || ashift > ASHIFT_MAX)) {
+				(void) fprintf(stderr, dgettext(TEXT_DOMAIN,
+				    "property 'ashift' value %" PRIu64 " is "
+				    "invalid.\n"), ashift);
```
It would be nice to provide the valid range in the error message.

loli10K replied:
I'll also update `zpool_valid_proplist()` error message adding "_only values between $ASHIFT_MIN and $ASHIFT_MAX are allowed_"

behlendorf commented on module/zfs/vdev.c:
```diff
@@ -1346,7 +1346,15 @@ vdev_open(vdev_t *vd)
 		if (vd->vdev_ashift == 0)
-			vd->vdev_ashift = ashift;
+			vd->vdev_ashift = ashift; /* use detected value */
+		else {
```
cstyle: Normally please add `{ }` to the `if` portion of this kind of conditional.  That said, in this case we should do the range change on `vd->vdev_ashift` regardless of how it was set.

behlendorf commented on cmd/zpool/zpool_vdev.c:
```diff
@@ -721,8 +722,18 @@ make_leaf_vdev(nvlist_t *props, const char *arg, uint64_t is_log)
+				(void) fprintf(stderr,
+				    gettext("property 'ashift' value %" PRIu64
+				    " is invalid: only values between %" PRId32
+				    " and %" PRId32 " are allowed.\n"),
```
Can we condense this slightly to fit in 80 character.  How about:

```
invalid 'ashift=X' property: only values between Y and Z are allowed.
```

behlendorf commented:
@loli10K thanks, LGTM.  I'll get this merged.
